concordium-smart-contract-testing/src/invocation/impls.rs

```diff
@@ -0,0 +1,1468 @@
+                }
+            }
+        }
+    }
+
+    /// Add two [`Self`] to create a new one.
+    fn add_delta(self, delta: AmountDelta) -> Self {
+        match delta {
+            AmountDelta::Positive(d) => self.add_amount(d),
+            AmountDelta::Negative(d) => self.subtract_amount(d),
+        }
+    }
+
+    /// Whether the [`Self`] is zero (either `+0` or `-0`).
+    fn is_zero(&self) -> bool {
+        match self {
+            AmountDelta::Positive(d) => d.micro_ccd == 0,
+            AmountDelta::Negative(d) => d.micro_ccd == 0,
+        }
+    }
+
+    /// Returns a new balance with the `AmountDelta` applied, or, an
+    /// error if `balance + self < 0`.
+    ///
+    /// Panics if an overflow occurs.
+    fn apply_to_balance(&self, balance: Amount) -> Result<Amount, UnderflowError> {
+        match self {
+            AmountDelta::Positive(d) => Ok(balance
+                .checked_add(*d)
+                .expect("Overflow occured when adding Amounts.")), // TODO: Should we return an
+            // error for this?
+            AmountDelta::Negative(d) => balance.checked_sub(*d).ok_or(UnderflowError),
+        }
+    }
+}
+
+impl ContractChanges {
+    /// Create a new `Self`. The original balance must be provided, all other
+    /// fields take on default values.
+    fn new(original_balance: Amount) -> Self {
+        Self {
+            modification_index:    0,
+            self_balance_delta:    AmountDelta::new(),
+            self_balance_original: original_balance,
+            state:                 None,
+            module:                None,
+        }
+    }
+
+    /// Get the current balance by adding the original balance and the balance
+    /// delta.
+    ///
+    /// **Preconditions:**
+    ///  - `balance_delta + original_balance` must be larger than `0`.
+    fn current_balance(&self) -> Amount {
+        self.self_balance_delta
+            .apply_to_balance(self.self_balance_original)
+            .expect("Precondition violation: invalid `balance_delta`.")
+    }
+}
+
+impl AccountChanges {
+    /// Get the current balance by adding the original balance and the balance
+    /// delta.
+    ///
+    /// **Preconditions:**
+    ///  - `balance_delta + original_balance` must be larger than `0`.
+    fn current_balance(&self) -> Amount {
+        self.balance_delta
+            .apply_to_balance(self.original_balance)
+            .expect("Precondition violation: invalid `balance_delta`.")
+    }
+}
+
+impl<'a> InvocationData<'a> {
+    /// Process a receive function until completion.
+    ///
+    /// **Preconditions**:
+    ///  - Contract instance exists in `invocation_handler.contracts`.
+    ///  - Account exists in `invocation_handler.accounts`.
+    fn process(
+        &mut self,
+        res: ExecResult<v1::ReceiveResult<artifact::CompiledFunction>>,
+    ) -> ExecResult<v1::ReceiveResult<artifact::CompiledFunction>> {
+        match res {
```

Comment:
I think you can avoid the outer match by using `?`.
It would same some indentation